compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -1206,38 +1206,54 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     fn suggest_range_for_iter(
         &self,
         tcx: TyCtxt<'tcx>,
-        source: SelfSource<'_>,
+        source: SelfSource<'tcx>,
         span: Span,
         item_name: Ident,
         ty_str: &str,
     ) -> bool {
         if let SelfSource::MethodCall(expr) = source {
             // avoid suggesting when the method name is not implemented for a `range`
-            // FIXME: any better solution?
-            let valid_methods = [
-                "rev",
-                "end",
-                "sum",
-                "take",
-                "start",
-                "end",
-                "is_empty",
-                "is_sorted",
-                "contains",
-                "start_bound",
-                "end_bound",
-            ];
+            let mut search_limit = 4;
+            for (_, parent) in tcx.hir().parent_iter(expr.hir_id) {
+                search_limit -= 1;
+                if search_limit == 0 {
+                    return false;
+                }
 
-            if !valid_methods.contains(&item_name.name.as_str()) {
-                return false;
-            }
-            for (hir_id, _parent) in tcx.hir().parent_iter(expr.hir_id) {
-                let parent = self.tcx.hir().find(hir_id);
-                if let Some(Node::Expr(parent_expr)) = parent &&
-                    is_range_literal(parent_expr) &&
-                    let hir::ExprKind::Struct(_, eps, _) = parent_expr.kind &&
-                    eps[1].expr.span.contains(expr.span) {
-                        debug!("parent_expr: {:?}", parent_expr);
+                if let Node::Expr(parent_expr) = parent &&
+                    is_range_literal(parent_expr) {
+                    let span_included = match parent_expr.kind {
+                            hir::ExprKind::Struct(_, eps, _) => eps.iter().any(|ep| ep.expr.span.contains(span)),
+                            // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
+                            hir::ExprKind::Call(ref func, ..) => func.span.contains(span),
+                            _ => false,
+                    };
+
+                    if !span_included {
+                        continue;
+                    }
+                    println!("parent_expr: {:?}", parent_expr);
+
+                    let range_def_id = self.tcx.lang_items().range_struct().unwrap();
+                    let range_ty = self.tcx.type_of(range_def_id);
```

Comment:
this is `Range<T>` which doesn't implement `Iterator`. You have to use `self.tcx.bound_type_of(range_def_id).subst(self.tcx, my_integer_type)`